### ISSUE TITLE
Bug: Error during 'sqlfluff lint': "NotImplementedError: Don't know what to do with negative indent balance (-1)."

### ISSUE THREAD

**Issue by barrywhart** [alanmcruickshank/sqlfluff#124]
This is an error I saw when running against the `master` branch at commit `4801f884ad87300eca2b66ded7e933feb2d8c34c`.

`test.sql`:
```
WITH
    all_upstream_matches AS (
        SELECT
            ROW_NUMBER() OVER
            (
                PARTITION BY low_business_type, low_size_label, low_gender_label, low_age_label
            ) AS rownum
)

SELECT
    count_correctly_substituted
```

Result:
```
$ sqlfluff lint test.sql
Traceback (most recent call last):
  File "/Users/bhart/.pyenv/versions/sqlfluff-3.6.4/bin/sqlfluff", line 11, in <module>
    load_entry_point('sqlfluff', 'console_scripts', 'sqlfluff')()
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/cli/commands.py", line 140, in lint
    result = lnt.lint_paths(paths, verbosity=verbose)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 622, in lint_paths
    result.add(self.lint_path(path, verbosity=verbosity, fix=fix))
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 609, in lint_path
    fix=fix, config=config))
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 552, in lint_string
    lerrs, _, _, _ = crawler.crawl(parsed)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 207, in crawl
    raw_stack=raw_stack, fix=fix, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 207, in crawl
    raw_stack=raw_stack, fix=fix, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 207, in crawl
    raw_stack=raw_stack, fix=fix, memory=memory)
  [Previous line repeated 3 more times]
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 177, in crawl
    raw_stack=raw_stack, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/std.py", line 1190, in _eval
    indent_balance))
NotImplementedError: Don't know what to do with negative indent balance (-1).
```

If I remove the `PARTITION BY` line, the error does not occur.

**Comment by alanmcruickshank** [comment]
Thanks @barrywhart - this is great feedback for the new indent work I've been doing. The detailed failure case is especially good. I'm already building it into my branch 😄 .

**Comment by alanmcruickshank** [comment]
This is fixed in #133
